File: output/python/tests/22_operations_basic.py
from typing import Any, Optional, List, Dict, Callable

class OperationsTestFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters


class OperationsTestFrameContext:
    def __init__(self, event: OperationsTestFrameEvent, default_return):
        self.event = event
        self._return = default_return
        self._data = {}


class OperationsTestCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'OperationsTestCompartment':
        c = OperationsTestCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class OperationsTest:
    def __init__(self):
        self._state_stack = []
        self._context_stack = []
        self.last_result = 0
        self.__compartment = OperationsTestCompartment("Ready")
        self.__next_compartment = None
        __frame_event = OperationsTestFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = OperationsTestFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = OperationsTestFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = OperationsTestFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def compute(self, a: int, b: int) -> int:
        __e = OperationsTestFrameEvent("compute", {"a": a, "b": b})
        __ctx = OperationsTestFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        return self._context_stack.pop()._return

    def get_last_result(self) -> int:
        __e = OperationsTestFrameEvent("get_last_result", None)
        __ctx = OperationsTestFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        return self._context_stack.pop()._return

    def _state_Ready(self, __e):
        if __e._message == "compute":
            a = __e._parameters["a"]
            b = __e._parameters["b"]
            # Use instance operations
            sum_val = self.add(a, b)
            prod_val = self.multiply(a, b)
            last_result = sum_val + prod_val
            self._context_stack[-1]._return = last_result
            return
        elif __e._message == "get_last_result":
            self._context_stack[-1]._return = last_result
            return

    def add(self, x: int, y: int) -> int:
        return x + y

    def multiply(self, x: int, y: int) -> int:
        return x * y

    @staticmethod
    def factorial(n: int) -> int:
        if n <= 1:
            return 1
        return n * OperationsTest.factorial(n - 1)

    @staticmethod
    def is_even(n: int) -> bool:
        return n % 2 == 0


def main():
    print("=== Test 22: Operations Basic ===")
    s = OperationsTest()

    # Test 1: Instance operations
    result = s.add(3, 4)
    assert result == 7, f"Expected 7, got {result}"
    print(f"1. add(3, 4) = {result}")

    result = s.multiply(3, 4)
    assert result == 12, f"Expected 12, got {result}"
    print(f"2. multiply(3, 4) = {result}")

    # Test 2: Operations used in handler
    result = s.compute(3, 4)
    # compute returns add(3,4) + multiply(3,4) = 7 + 12 = 19
    assert result == 19, f"Expected 19, got {result}"
    print(f"3. compute(3, 4) = {result}")

    # Test 3: Static operations
    result = OperationsTest.factorial(5)
    assert result == 120, f"Expected 120, got {result}"
    print(f"4. factorial(5) = {result}")

    result = OperationsTest.is_even(4)
    assert result == True, f"Expected True, got {result}"
    print(f"5. is_even(4) = {result}")

    result = OperationsTest.is_even(7)
    assert result == False, f"Expected False, got {result}"
    print(f"6. is_even(7) = {result}")

    # Test 4: Static can also be called on instance
    result = s.factorial(4)
    assert result == 24, f"Expected 24, got {result}"
    print(f"7. instance.factorial(4) = {result}")

    print("PASS: Operations basic works correctly")

if __name__ == '__main__':
    main()

File: output/javascript/tests/22_operations_basic.js

export class OperationsTestFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class OperationsTestFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class OperationsTestCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new OperationsTestCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class OperationsTest {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;
    last_result = 0;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new OperationsTestCompartment("Ready");
        this.__next_compartment = null;
        const __frame_event = new OperationsTestFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new OperationsTestFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new OperationsTestFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new OperationsTestFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    compute(a, b) {
        const __e = new OperationsTestFrameEvent("compute", {"a": a, "b": b});
        const __ctx = new OperationsTestFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    get_last_result() {
        const __e = new OperationsTestFrameEvent("get_last_result", null);
        const __ctx = new OperationsTestFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    _state_Ready(__e) {
        if (__e._message === "compute") {
            const a = __e._parameters?.["a"];
            const b = __e._parameters?.["b"];
            // Use instance operations
            const sum_val = this.add(a, b);
            const prod_val = this.multiply(a, b);
            const last_result = sum_val + prod_val;
            this._context_stack[this._context_stack.length - 1]._return = last_result;
            return;;
        } else if (__e._message === "get_last_result") {
            this._context_stack[this._context_stack.length - 1]._return = this.last_result;
            return;;
        }
    }

    add(x, y) {
                    return x + y;
    }

    multiply(x, y) {
                    return x * y;
    }

    static factorial(n) {
                    if (n <= 1) {
                        return 1;
                    }
                    return n * OperationsTest.factorial(n - 1);
    }

    static is_even(n) {
                    return n % 2 === 0;
    }
}

function main() {
    console.log("=== Test 22: Operations Basic (JavaScript) ===");
    const s = new OperationsTest();

    // Test 1: Instance operations
    let result = s.add(3, 4);
    if (result !== 7) {
        throw new Error(`Expected 7, got ${result}`);
    }
    console.log(`1. add(3, 4) = ${result}`);

    result = s.multiply(3, 4);
    if (result !== 12) {
        throw new Error(`Expected 12, got ${result}`);
    }
    console.log(`2. multiply(3, 4) = ${result}`);

    // Test 2: Operations used in handler
    result = s.compute(3, 4);
    // compute returns add(3,4) + multiply(3,4) = 7 + 12 = 19
    if (result !== 19) {
        throw new Error(`Expected 19, got ${result}`);
    }
    console.log(`3. compute(3, 4) = ${result}`);

    // Test 3: Static operations
    result = OperationsTest.factorial(5);
    if (result !== 120) {
        throw new Error(`Expected 120, got ${result}`);
    }
    console.log(`4. factorial(5) = ${result}`);

    let is_even = OperationsTest.is_even(4);
    if (is_even !== true) {
        throw new Error(`Expected true, got ${is_even}`);
    }
    console.log(`5. is_even(4) = ${is_even}`);

    is_even = OperationsTest.is_even(7);
    if (is_even !== false) {
        throw new Error(`Expected false, got ${is_even}`);
    }
    console.log(`6. is_even(7) = ${is_even}`);

    // Test 4: Static via class
    result = OperationsTest.factorial(4);
    if (result !== 24) {
        throw new Error(`Expected 24, got ${result}`);
    }
    console.log(`7. OperationsTest.factorial(4) = ${result}`);

    console.log("PASS: Operations basic works correctly");
}

main();
